### DISCUSSION TITLE
[TypeScript] How to get `command` from context?

### DISCUSSION THREAD

**Discussion by hashfi-motion** [telegraf/telegraf#1952] [Q&A] [answered]
As the title said, how do I get `command` from context? I tried `deunionize` the context but TypeScript still can't find it

![command](https://github.com/telegraf/telegraf/assets/129708704/ec6c34ca-fdb2-431d-84ec-ceb86cb65964)
![error](https://github.com/telegraf/telegraf/assets/129708704/a1514ccc-ecd4-4c82-9c60-c6916bd06dae)

**Comment by MKRhere** [comment]
`ctx.command` is only available in `bot.command` and `bot.start`. It doesn't exist in `Context` by itself.

**Reply by hashfi-motion** [reply]
Hmm, how do I access it then? My code is something like below:

```typescript
	bot.command(/.*/, _before);
	bot.action(/.*/, _before);
	
	//...
	_before = async (cx: ContextSession) => {
		//...
	};
```

`ContextSession` basically just Context with some `session` I add

**Reply by MKRhere** [reply]
```TS
interface ContextSession <U extends Update = Update> extends Context<U> {
	session: MySessionType;
}

_before = async (ctx: ContextSession<Update.MessageUpdate<Message.TextMessage>> & { command: string }) {
	// ...
}
```

>Note: The & part is internally `CommandContextExtn` and is not exported. In the upcoming `4.16.0` release, I'm exposing it as `Convenience.CommandContextExtn`. The following will be made possible:
>
>	```TS
>	import type { Convenience } from "telegraf/types";
>	
>	// ContextSession<Update.MessageUpdate<Message.TextMessage>> & CommandContextExtn
>	```

**Reply by MKRhere** [reply, marked as answer]
Also realise `bot.action(/.*/, _before)` will NOT have `ctx.command`. It's probably better you extract the common logic out of `_before`, then call it like so:

```TS
bot.command(/.*/, ctx => {
  return commonHandler(ctx.command, ...);
});

bot.action(/.*/, ctx => {
	return commonHandler(ctx.match, ...);
});
```

**Reply by hashfi-motion** [reply]
Thanks! I guess I'm gonna go with the second approach, the first approach just look ugly and long 😅
